### PR TITLE
fix(#8566): adds sentinel transition retry cache

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Adds a cache of 1000 entries (lifo) that tracks how many times Sentinel tried to run transition over a specific change (id + rev). Once the retry count is over 5, the change is skipped and the sentinel queue goes past it. 

medic/cht-core#8566

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8539-adds-sentinel-transition-retry-cache/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8539-adds-sentinel-transition-retry-cache/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8539-adds-sentinel-transition-retry-cache/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by garethbowen** [comment]
@dianabarsan Is this ready for review?

**Comment by dianabarsan** [comment]
I _think_ so, @garethbowen . Would you mind?

**Review by garethbowen — APPROVED** [review]
One optional readability suggestion inline.

File: sentinel/src/lib/change-retry-history.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+const historyKeys = [];
+const history = {};
+
+const getKey = (change) => change && `${change.id}${change.changes && change.changes[0] && change.changes[0].rev}`;
+
+const add = (change) => {
+  const key = getKey(change);
+  if (!key) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply sentinel/src/lib/change-retry-history.js]: refactored :)

**Pull request merged**